### PR TITLE
work towards chalkify-ing the engine

### PR DESCRIPTION
**Pull request by nikomatsakis**
This work towards creating a "all program clauses needed for this goal" query

r? @scalexm

### CODE REVIEW

**Comment by nikomatsakis** [comment]
D'oh, have to rebase =)

**Comment by nikomatsakis** [comment]
rebased

**Comment by kennytm** [comment]
Legit. The chalk rules should better be sorted when dumped out...

```
[00:48:07] ---- [ui] ui/chalkify/lower_env1.rs stdout ----
[00:48:07] 	diff of stderr:
[00:48:07] 
[00:48:07] 14	LL | #[rustc_dump_env_program_clauses] //~ ERROR program clause dump
[00:48:07] 15	   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:48:07] 16	   |
[00:48:07] +	   = note: Implemented(Self: std::marker::Sized) :- FromEnv(Self: std::marker::Sized).
[00:48:07] 17	   = note: Implemented(Self: Bar) :- FromEnv(Self: Bar).
[00:48:07] 18	   = note: FromEnv(Self: Bar) :- FromEnv(Self: Bar).
[00:48:07] 19	   = note: FromEnv(Self: Foo) :- FromEnv(Self: Bar).
[00:48:07] 
[00:48:07] -	   = note: Implemented(Self: std::marker::Sized) :- FromEnv(Self: std::marker::Sized).
[00:48:07] 21	   = note: Implemented(Self: Foo) :- FromEnv(Self: Foo).
[00:48:07] 22	
[00:48:07] 23	error: aborting due to 2 previous errors
```

File: src/librustc/hir/map/definitions.rs [hunk 1]
```diff
@@ -358,6 +361,12 @@ pub enum DefPathData {
     // Different kinds of items and item-like things:
     /// An impl
     Impl,
+    /// A trait
+    Trait(InternedString),
+    /// An associated type **declaration** (i.e., in a trait)
+    AssocTypeInTrait(InternedString),
+    /// An associated type **value** (i.e., in an impl)
+    AssocTypeInImpl(InternedString),
```
Comment (eddyb) [inline src/librustc/hir/map/definitions.rs]: Why was this split done? Also, why "declaration vs value" and not "declaration vs definition"?

File: src/librustc_traits/lowering.rs [hunk 2]
```diff
@@ -104,44 +113,88 @@ impl<'tcx> IntoFromEnvGoal for DomainGoal<'tcx> {
...
-            }
-        },
+crate fn program_clauses_for<'a, 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    def_id: DefId,
+) -> Clauses<'tcx> {
+    match tcx.def_key(def_id).disambiguated_data.data {
+        DefPathData::Trait(_) => program_clauses_for_trait(tcx, def_id),
+        DefPathData::Impl => program_clauses_for_impl(tcx, def_id),
+        DefPathData::AssocTypeInImpl(..) => program_clauses_for_associated_type_value(tcx, def_id),
```
Comment (eddyb) [inline src/librustc_traits/lowering.rs]: I'm not sure I approve of using `DefPathData` outside of path-specific logic.
Maybe we need a new enum that's something in between `DefPathData` and `Def`?
"`DefKind`" comes to mind as a possible name.

cc @oli-obk @varkor

Comment (oli-obk) [reply src/librustc_traits/lowering.rs]: we already have https://github.com/rust-lang/rust/blob/059e6a6f57f4e80d527a3cd8a8afe7f51f01af8e/src/librustc/hir/def.rs#L46

Comment (eddyb) [reply src/librustc_traits/lowering.rs]: By "something in between `DefPathData` and `Def`" I meant "not `Def` itself" - ideally, a plain enum with no `DefId`s.

Comment (eddyb) [reply src/librustc_traits/lowering.rs]: Oh and `Def` should probably be renamed to `Res`(`olution`), as it's too specific to name resolution to deserve a more general name.

**Comment by oli-obk** [comment]
So... We could make `Def` a `struct Def { kind: DefKind, id: Option<DefId> }` ?

**Comment by varkor** [comment]
The variants of `Def` aren't quite so homogeneous (e.g. `SelfTy(Option<DefId>, Option<DefId>)`, `StructCtor(DefId, CtorKind)`, etc.), so I'm not sure it'd make such a clean refactoring. It would probably end up duplicating the `Def` enum and stripping out the unnecessary data.

**Comment by eddyb** [comment]
@oli-obk @varkor My preferred solution would be to have `Res::Def(DefKind, DefId)` (if we even want to bother with duplicating the `DefKind` there instead of having a map/query from `DefId` to `DefKind`)

That is, `Def` would become `Res`, it would keep all of its variants which *don't* just refer to definitions (in the sense of having a `DefId`), e.g. `SelfTy`, `Local`, etc. and gain a new `Def(DefKind, DefId)` variant.

**Pull request merged**